### DISCUSSION TITLE
Cannot use Chrome browser behind proxy

### DISCUSSION THREAD

**Discussion by eqMFqfFd** [php-webdriver/php-webdriver#876] [Q&A] [answered]
### Bug description
When I try to start up the Chrome broser using proxy, a `WebDriverException` is thrown.

### How could the issue be reproduced
1.   Run the script.
2. See the error.

Test 1
```php
$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability(WebDriverCapabilityType::PROXY, [
	"proxyType" => "manual",
	"httpProxy" => "60.234.24.162:3128"
]);
$driver = RemoteWebDriver::create("http://localhost:4444/", $desiredCapabilities);
$driver->get("https://www.google.com/");
```

Test 2
```php
$driver = RemoteWebDriver::create("http://localhost:4444/", null, null, null, "60.234.24.162", 3128);
$driver->get("https://www.google.com/");
```

Error
```
PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\WebDriverException
: JSON decoding of remote response failed.
```

### Expected behavior
I should be able to successfully start the browser behind a proxy.

### Details

* Php-webdriver version: 1.10.0
* PHP version: 8.0.1 (cli)
* How do you start the browser: PHP script
~~* Selenium server version:~~
* Operating system:  Ubuntu 18.0.4.5
* Browser used + version: ChromeDriver 89.0.4389.90
* Browser driver (chromedriver/geckodriver...) version: chromedriver

### Additional context
I tried starting browser by passing the proxy address and port to the factory. I also tried using browser capabilities. None of them work.

**Comment by andrewnicols** [comment]
Hi @kenygamer ,

It looks like there's a few things going on here:
1. You're using an webdriver URL of `http://localhost:4444/` but you say that you're using Chromedriver without Selenium. The default URL is `http://localhost:9515/` for chromedriver, or `http://localhost:4444/wd/hub` via Selenium. Just confirm that you are either running chromedriver on the non-default port, or that you have the correct URL.
2. Your option 1: You're setting an http proxy, but accessing an https site. You have not configured an sslProxy setting so everything is local.
2. For your option 2: The URL for chromedriver is local to the _proxy server_ - so that `http://localhost:4444/` means `localhost` on the SOCKS proxy you're using. When provided to the factory method it is used as a proxy to reach your webdriver host. It is not the proxy provided to the browser.

If I copy your Option 1 code and add the sslProxy, it does work as expected.

Note: You have an open relay proxy. That's generally speaking a really bad thing. I suggest you close that ASAP.

**Reply by andrewnicols** [reply]
The following script demonstrates correct working http proxy when the sslProxy is specified:

```
<?php

namespace Facebook\WebDriver;

use DOMDocument;
use Facebook\WebDriver\Remote\DesiredCapabilities;
use Facebook\WebDriver\Remote\RemoteWebDriver;
use Facebook\WebDriver\Remote\WebDriverCapabilityType;

require_once('vendor/autoload.php');

$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability(WebDriverCapabilityType::PROXY, [
    "proxyType" => "manual",
    "httpProxy" => "60.234.24.162:3128"
]);
$desiredCapabilities->setCapability(WebDriverCapabilityType::PROXY, [
    "proxyType" => "manual",
    "sslProxy" => "60.234.24.162:3128"
]);
$driver = RemoteWebDriver::create("http://localhost:9515/", $desiredCapabilities);
$driver->get("https://api.ipify.org");
$source = $driver->getPageSource();
$driver->quit();

$dom = new DOMDocument();
$dom->loadHTML($source);
error_log($dom->textContent);
```

**Comment by eqMFqfFd** [comment]
@andrewnicols Thanks. I have tried your proposed solutions and I have sticked with the first one. I see success with local tests, but some sites are apparently blocking me.

When I try to fetch them with the `wget` CLI tool, I can successfully get the page results, but when I use php-webdriver (behind/not behind proxy), I get a timeout error.

After inspecting the headers, I was able to confirm that the browser name is set to Chrome headless. 

```
[SERVER_SOFTWARE] => Apache/2.4.29 (Ubuntu)
Mozilla/5.0 (X11; Linux x86_64) AppleWebKit/537.36 (KHTML, like Gecko) HeadlessChrome/89.0.4389.90 Safari/537.36
```

Is there a way to hide the browser name, or to remove the website block?

**Reply by andrewnicols** [reply]
I don't believe so. The User Agent can be controlled (as in the browser can be controlled), but the User Agent String cannot.
There is a languishing feature request at https://github.com/w3c/webdriver/issues/1459 but I can't see it changing any time soon.

**Reply by eqMFqfFd** [reply]
Is there a way to push this feature suggestion or some provisional workaround?

**Reply by OndraM** [reply]
I'd surprised if someone actually use the User Agent to block the requests. Can you provide the website URL?

BTW to change user agent, there is possibility using Chrome DevTools protocol: https://github.com/php-webdriver/php-webdriver/wiki/Chrome#change-user-agent

**Reply by eqMFqfFd** [reply]
> I'd surprised if someone actually use the User Agent to block the requests. Can you provide the website URL?
> 
> BTW to change user agent, there is possibility using Chrome DevTools protocol: https://github.com/php-webdriver/php-webdriver/wiki/Chrome#change-user-agent

Thank you. Also, can [this](https://github.com/php-webdriver/php-webdriver/wiki/Chrome#set-custom-request-header) change other headers? (Like `Accept-Encoding`, etc.)

**Reply by OndraM** [reply]
I can see there is method especially for this in CDP, so you can give it a try.

https://chromedevtools.github.io/devtools-protocol/tot/Network/#method-setAcceptedEncodings

**Reply by eqMFqfFd** [reply]
The following code does not add **any** of the specified custom headers:

**Code to reproduce:**

```php
$devTools = new ChromeDevToolsDriver($driver);
$devTools->execute("Network.enable");
$devTools->execute("Network.setExtraHTTPHeaders", [
	"headers" => (object) [
	    'Accept-Language' => 'en-en',
	    'Connection' => 'keep-alive',
	    'Upgrade-Insecure-Requests' => '1',
	     'Referer' => 'https://www.google.com/',
	     'User-Agent' => 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_6) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/14.0.1 Safari/605.1.15'
        ]
]);
```

**Output:**

```
<html><head></head><body></body> </html>
```

**Reply by eqMFqfFd** [reply]
@OndraM

**Reply by OndraM** [reply]
I don't understand what do you mean by the "Output" part.

We cannot affect behavior of CDP. The `Network.setExtraHTTPHeaders` obviously allow to change only some of the headers - its not true it does not change "any" of headers. I just tested this example and at least "Accept-Language" and "User-Agent" are definitely changed.

And as mentioned, this is completely up to Chrome - so if some CDP behavior does not fits you, please refer to CDP documentation or user forums.

**Comment by eqMFqfFd** [comment]
@andrewnicols I have come across a new issue with my experience using proxy with php-webdriver.

One of my proxies needs authentication. I went crazy looking for how to authenticate proxy, until I found this article on the wiki:

[HowTo Work with proxy](https://github.com/php-webdriver/php-webdriver/wiki/HowTo-Work-with-proxy)

The example code throws the following error on the console:

```
Fatal error: Uncaught Facebook\WebDriver\Exception\UnknownErrorException: unknown error: failed to wait for extension background page to load: chrome-
extension://ejehnedamdfklallemmmgajfiokdmcab/_generated_background_page.html
```

According to [this article](https://stackoverflow.com/a/62346996) on Stackoverflow, it is not possible to use authenticated proxy with Chrome headless.

I have disabled the `--headless` option, which throws the following error:

```
Fatal error: Uncaught Facebook\WebDriver\Exception\UnknownErrorException: unknown error: Chrome failed to start: exited abnormally.
```

What am I doing wrong?

**Reply by OndraM** [reply]
Why don't you provide credentials in the proxy URL?

```php
[
    'httpProxy' => 'user:password@my.own.proxy:8118',
    'sslProxy' => 'user:password@my.own.proxy:8118',
]
```

**Reply by eqMFqfFd** [reply]
That didn't work out.

The tests I just ran include:

Fake proxy used: `google.com:3030`
Live proxy: `user:password@real.proxy:3030`

- Live proxy using PHP curl - OK
- Live proxy using php-webdriver - The page still loads, using the IP of the calling script.
- Fake proxy using php-webdriver - The proxy is silently ignored.

**Reply by OndraM** [reply]
Can you provide some reproducible script? At least the capabilities configuration and how you create webdriver instance.

**Reply by eqMFqfFd** [reply]
@OndraM This is my configuration:

Chromedriver options: `--headless, --disable-dev-shm-usage, --no-sandbox`
Chromedriver experimental options: `excludeSwitches, enable-automation`

**Reply by OndraM** [reply]
@kenygamer I need to see the PHP code you use to create capabilities (including the full code how you setup the proxy) and how you start the $driver instance.

**Reply by eqMFqfFd** [reply]
Full PHP code:
```php
$proxy = "[redacted username]:[redacted password]@[redacted ip]:[redacted port]";
$serverUrl = "http://localhost:4444/";
$desiredCapabilities = DesiredCapabilities::chrome();
$options = new ChromeOptions();
$options->addArguments(["--headless",  "--disable-dev-shm-usage", "--no-sandbox", "--test-type", "--disable-extensions", "disable-infobars", "--disable-notifications", "enable-automation", "--disable-popup-blocking", "start-maximized"]);
$options->setExperimentalOption("excludeSwitches", ["enable-automation"]);
$options->setExperimentalOption("useAutomationExtension", false);
$desiredCapabilities->setCapability(WebDriverCapabilityType::PROXY, [
	"proxyType" => "manual",
	"httpProxy" => $proxy,
	"sslProxy" => $proxy
]);
		
$desiredCapabilities->setCapability(ChromeOptions::CAPABILITY_W3C, $options);
$this->driver = RemoteWebDriver::create($serverUrl, $desiredCapabilities);
$this->driver->get("http://localhost/test_proxy.php");
print_r($this->driver->getPageSource()); //See that IP does not change
```

**Reply by eqMFqfFd** [reply]
@OndraM

**Reply by OndraM** [reply]
It does work for me. If I provide some not-working proxy URL, it throws error as expected:

`PHP Fatal error:  Uncaught Facebook\WebDriver\Exception\UnknownErrorException: unknown error: net::ERR_CONNECTION_REFUSED`

Other simpler example:
```php
$host = 'http://localhost:4444';
$proxy = 'foo.bar:300';
$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability(
    WebDriverCapabilityType::PROXY,
    [
        'proxyType' => 'manual',
        'httpProxy' => $proxy,
        'sslProxy' => $proxy,
    ]
);
$driver = RemoteWebDriver::create($host, $desiredCapabilities);
$driver->get('https://google.com/');

```

**Reply by eqMFqfFd** [reply]
In my case, php-webdriver does not throw any exceptions when I try a non-functional proxy. And when I try an open proxy, it works, but an authenticated proxy will silently fail.

**Reply by OndraM** [reply]
I just tried with authenticated proxy, it works properly.

What I tried:

1. Create local proxy in Docker using [mitmproxy](https://mitmproxy.org/). It requires authentication using `user` username and `pass` password.
```sh
docker run --rm -it -p 8080:8080 -p 127.0.0.1:8081:8081 mitmproxy/mitmproxy mitmweb --web-host 0.0.0.0 --set proxyauth=user:pass
```

2. Start local chromedriver
```sh
chromedriver --port=4444
```

3. Create PHP script `proxy-example.php`
```php
<?php

use Facebook\WebDriver\Remote\DesiredCapabilities;
use Facebook\WebDriver\Remote\RemoteWebDriver;
use Facebook\WebDriver\Remote\WebDriverCapabilityType;

require_once('vendor/autoload.php');

$proxy = 'user:pass@localhost:8080';

$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability(
    WebDriverCapabilityType::PROXY,
    [
        'proxyType' => 'manual',
        'httpProxy' => $proxy,
        'sslProxy' => $proxy,
    ]
);
$driver = RemoteWebDriver::create('http://localhost:4444', $desiredCapabilities);

$driver->get('http://neverssl.com/');
```

4. Run PHP script
```sh
php proxy-example.php
```

5. The page is loaded via authenticated  proxy.

**Reply by eqMFqfFd** [reply]
I can confirm that this does not work with remote proxies. I can also confirm that this is an issue with php-webdriver, as the C# bindings work perfectly.

**Reply by OndraM** [reply]
There is IMHO no reason why this would work with "local proxy" and not "remote proxy"...

Can you have look at chromedriver output? If you start chromedriver like this:
```
chromedriver --log-level=INFO --port=4444
```
it will produce more verbose output. Then start the script where you initialize the proxy and where you try to navigate to the webpage.

Could you provide the full output from chromedriver? (Don't forget to redact the passwords there).

**Reply by eqMFqfFd** [reply]
https://gist.github.com/kenygamer/c6027ff603476c011351785efbdc2e17

The attached logs occurs when the page loads, but without any proxied connection (the address is the origin IP)

**Reply by OndraM** [reply]
Why do you pass `--proxy-auto-detect` option to Chrome? From what I tried, it makes the proxy capabilities being ignored.

Also this option was not part of the full PHP code you provided [above](https://github.com/php-webdriver/php-webdriver/discussions/876#discussioncomment-581533)...

If this won't help, try to debug the issue - remove all other chrome options to make minimal reproducible case. And then provide the full PHP code and chromedriver log here, please.

**Reply by eqMFqfFd** [reply]
I cannot just "remove" all the other chrome options. I am using `--headless`, because I don't have a VPS with GUI, `--no-sandbox` to fix startup issues, or `--disable-dev-shm-usage` to prevent chrome automation detection.

This is the list of chrome options that I am now testing with.
```
--headless, --disable-dev-shm-usage, --no-sandbox, --test-type, --disable-extensions, disable-infobars, --disable-notifications, --enable-automation, --disable-popup-blocking, --start-maximized
```

Your proposed solution doesn't fix my issue.

**Reply by OndraM** [reply]
Further investigation shows the proxy authentication according to W3C WebDriver specification is not fully implemented in neither chrome (it probablby ignores the credentials) nor firefox.

See issue https://github.com/php-webdriver/php-webdriver/issues/885 for more information and and [wiki for possible workaround](https://github.com/php-webdriver/php-webdriver/wiki/HowTo-Work-with-proxy#proxy-with-authentication) using mitmproxy.

**Comment by eqMFqfFd** [comment]


**Reply by OndraM** [reply]
Please don't ping me like this. This is community discussion and nor I nor anyone else have an obligation to respond.

And this is not related to this topic ("Cannot use Chrome browser behind proxy"). If you need, you can open a new topic in discussion.

**Comment by OndraM** [comment, marked as answer]
For the record, working example **for unauthenticated proxy**:

Start chromedriver
```sh
chromedriver --port=4444
```

Execute script:
```php
$host = 'http://localhost:4444';
$proxy = 'my.proxy.url:1234';

$desiredCapabilities = DesiredCapabilities::chrome();
$desiredCapabilities->setCapability(
    WebDriverCapabilityType::PROXY,
    [
        'proxyType' => 'manual',
        'httpProxy' => $proxy,
        'sslProxy' => $proxy,
    ]
);
$driver = RemoteWebDriver::create($host, $desiredCapabilities);
$driver->get('https://google.com/'); // will be accessed through proxy
```

**For proxy requiring authentication** you can use workaround [described in our wiki](https://github.com/php-webdriver/php-webdriver/wiki/HowTo-Work-with-proxy#proxy-with-authentication).
